### PR TITLE
Forward sats to user: text trimming and menu click-out UX improvements

### PR DESCRIPTION
**Pull request by rleed**
Approach: on input, check for leading/trailing spaces and the leading @ symbol on the username (e.g. when pasting a username with extra characters) and strip them. This has the effect of also restricting the user from typing such characters too.

While working on this, I discovered the drop-down suggestions menu was too sticky; it covered the post button and virtually obliged the user to pick a name in the list (even if not necessary because it was pasted in, e.g.) so I added logic to hide the menu when the user clicks out or tabs away.

(issue #185)

### CODE REVIEW

File: components/form.js [hunk 1]
```diff
@@ -375,6 +386,11 @@ export function InputUserSuggest ({ label, groupClassName, ...props }) {
                 setOValue(v.name)
                 setSuggestions(INITIAL_SUGGESTIONS)
               }}
+              onBlur={(e) => {
+                if (!document.getElementsByClassName(groupClassName)[0].contains(e.relatedTarget)) {
+                  setSuggestions(INITIAL_SUGGESTIONS)
+                }
+              }}
```
Comment (ekzyis) [inline components/form.js:393]: I can see that you need the `if` check in the input `onBlur` listener to be able to still select items. The `onClick` listener then seems to close the dropdown.

But do you need _this_ `onBlur` listener here? It seemed to work without this.

Also, can't this be solved in a more "react-y way"; without using `document.getElementsByClassName` For example with `useRef`? I didn't try yet though.

**Review by ekzyis — COMMENTED** [review]
It's hard to test this properly since I think `<InputUserSuggest>` in general is buggy. See #409. For example, already on 41463d7, one can't reselect users. Maybe there is more broken stuff.

But I noticed that with these changes, the icon if the input is valid no longer appears:

https://github.com/stackernews/stacker.news/assets/27162016/196f01b3-5707-4cc7-a4c3-d6f5e076db86

@rleed, can you confirm on your end?

I encountered strange things with this input with and without your changes while testing so I am not sure if your changes actually cause this.

I was also wondering why we can't use the input property [`pattern`](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/pattern) for this instead of the new custom prop `autoRestrict`. When I set it, it doesn't seem to have any effect even though I can see it set in the raw HTML. It is at least mentioned in the validation chapter of Formik: https://formik.org/docs/tutorial#validation

**Comment by rleed** [comment]
Re: testing, yes, the validation seems buggy or at least quirky and I also noticed that reselecting isn't recognized. This PR was intended as an incremental improvement...we could take a closer look at those other oddities in another PR.

I noticed the validation didn't happen till attempting to submit the form, and I seem to remember seeing something in the code to deliberately disable validation initially... I'll check that, because I didn't think I changed that behavior.

Using pattern would only give a validation signal but would not do anything to strip the redundant characters when pasting a username. This was considered as a possible behavior in the issue discussion, but IMO it doesn't make for a smooth UX, which is why I ended up restricting the input instead.

File: components/form.js (hunk 1 above)
Comment (rleed) [reply components/form.js:393]: The simplest way to think about it is that if the focus leaves the whole group (inclusive of the input box and the suggested items) then and only then do we want to hide the menu. The blur can be triggered by clicking or by forward/backward tabbing. So the blur handler hides the menu IF the new focused element is anything outside the group. And for that, all elements in the group that can get focus have to have the blur function.

I wanted to use useRef (commonly used for focus-handling use cases) but unfortunately useRef is no longer supported on function components it seems.

**Comment by rleed** [comment]
> But I noticed that with these changes, the icon if the input is valid no longer appears:

Good catch, thanks. (My blur handlers were overriding Formik's behavior.)

**Comment by rleed** [comment]
So if all are ok with my answers to @ekzyis, I think this one is ready

**Comment by huumn** [comment]
Thanks so much! Please keep in the mind the following feedback as you continue to submit prs.

I ended up removing the blur handler because while it's nice to have, it was a bit of hack and hard to follow. re: useRef. If you want to forward a ref from a parent to a child, you wrap the child in [`forwardRef`](https://react.dev/reference/react/forwardRef).

I also updated the change to reuse the code and patterns that were already present instead of writing new ones.
